### PR TITLE
Implement snapshot iterator

### PR DESCRIPTION
**Pull request by alyakimenko**
### Description

This patch seeks to implement a snapshot iterator.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-neo4j/pulls) for the same update/change.
- [ ] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: source/iterator/errors.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	// and this process fails.
+	errConvertRawNode = errors.New("cannot convert raw element to dbtype node")
+
+	// errConvertRawRelationship occurs when trying to convert a raw element to [dbtype.Relationship]
+	// and this process fails.
+	errConvertRawRelationship = errors.New("cannot convert raw element to dbtype relationship")
+
+	// neo4jNoMoreRecordsErrorMessage is a message
+	// that Neo4j returns when it cannot find records.
+	neo4jNoMoreRecordsErrorMessage = "Result contains no more records"
```
Comment (maksenius) [inline source/iterator/errors.go:38]: As for me Result contains no more records looks not clear. Maybe we can rename it to  "Result without records" or something similar

File: source/iterator/position.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	modeSnapshot positionMode = "snapshot"
+	modeCDC      positionMode = "cdc"
+)
+
+// position is an iterator position.
+type position struct {
+	Mode positionMode `json:"mode"`
+	// Element is a value of the last processed element by the snapshot capture.
+	// This value is used if the mode is snapshot.
+	Element any `json:"element"`
```
Comment (maksenius) [inline source/iterator/position.go:38]: What do you think about rename it to `Value` or `LastProccesedValue` ?

File: source/iterator/snapshot.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	srcPlaceholder                    = "src"
+	trgtPlaceholder                   = "trgt"
+
+	// relationship payload-specific fields.
+	sourceNodeField = "sourceNode"
+	targetNodeField = "targetNode"
+)
+
+// Snapshot implements a snapshot logic for the connector.
+type Snapshot struct {
```
Comment (maksenius) [inline source/iterator/snapshot.go:53]: Could we add short info about fields in comments ?

File: source/iterator/snapshot.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+	driver                   neo4j.DriverWithContext
+	orderingProperty         string
+	keyProperties            []string
+	orderingPropertyMaxValue any
+	entityType               config.EntityType
+	entityLabels             string
+	batchSize                int
+	databaseName             string
+	position                 *position
+	// use a channel here, as it's easier to work with it as a queue
```
Comment (maksenius) [inline source/iterator/snapshot.go:63]: please add short description about this channel

File: source/iterator/snapshot.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+	if err != nil && !errors.Is(err, errNilSDKPosition) {
+		return nil, fmt.Errorf("parse position: %w", err)
+	}
+
+	var (
+		orderingPropertyMaxValue any
+		// join entity labels here to not do this for each individual element
+		entityLabels = strings.Join(params.EntityLabels, ":")
+	)
+	switch pos := position; {
```
Comment (maksenius) [inline source/iterator/snapshot.go:91]: I think we can use if condition here. Or do we want to expand this switch in future?

File: source/iterator/snapshot.go [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+			if !ok {
+				return sdk.Record{}, fmt.Errorf("payload doesn't contain %q property", keyProperty)
+			}
+
+			key[keyProperty] = keyPropertyValue
+		}
+
+		// construct the metadata
+		metadata := make(sdk.Metadata)
+		metadata.SetCreatedAt(time.Now())
```
Comment (maksenius) [inline source/iterator/snapshot.go:167]: What do you think to add entity name to metadata?

File: source/iterator/snapshot.go [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+	position, err := parsePosition(params.Position)
+	if err != nil && !errors.Is(err, errNilSDKPosition) {
+		return nil, fmt.Errorf("parse position: %w", err)
+	}
+
+	var (
+		orderingPropertyMaxValue any
+		// join entity labels here to not do this for each individual element
+		entityLabels = strings.Join(params.EntityLabels, ":")
+	)
```
Comment (maksenius) [inline source/iterator/snapshot.go:90]: could you please ad breakline

File: source/iterator/errors.go (hunk 1 above)
Comment (alyakimenko) [reply source/iterator/errors.go:38]: It's Neo4j internal error message. We cannot change it.

File: source/iterator/position.go (hunk 2 above)
Comment (alyakimenko) [reply source/iterator/position.go]: Yeah, sure, `LastProcessedValue` would work fine. I've renamed it.

File: source/iterator/snapshot.go (hunk 3 above)
Comment (alyakimenko) [reply source/iterator/snapshot.go:53]: Most of the fields (except `driver`, `position`, and `records`) are from the configuration. These fields have descriptions there. Do I need to copy and paste descriptions from the configuration to this structure?

File: source/iterator/snapshot.go (hunk 4 above)
Comment (alyakimenko) [reply source/iterator/snapshot.go]: I've added a description to this field.

File: source/iterator/snapshot.go (hunk 5 above)
Comment (alyakimenko) [reply source/iterator/snapshot.go:93]: In this case, we'll need the `else` clause, which is a bit harder to read, in my opinion.

For comparison:
<table>
<tr>
<th><code>if/else</code></th>
<th><code>switch</code></th>
</tr>
<tr>
<td>

```go
if position != nil && position.MaxElement != nil {
	orderingPropertyMaxValue = position.MaxElement
} else {
	orderingPropertyMaxValue, err = getMaxPropertyValue(
		ctx, params.Driver,
		params.DatabaseName, entityLabels, params.OrderingProperty,
		params.EntityType,
	)
	if err != nil && !errors.Is(err, errNoElements) {
		return nil, fmt.Errorf("get ordering property max value: %w", err)
	}
}
```
</td>
<td>

```go
switch pos := position; {
case pos != nil && position.MaxElement != nil:
	orderingPropertyMaxValue = position.MaxElement

default:
	orderingPropertyMaxValue, err = getMaxPropertyValue(
		ctx, params.Driver,
		params.DatabaseName, entityLabels, params.OrderingProperty,
		params.EntityType,
	)
	if err != nil && !errors.Is(err, errNoElements) {
		return nil, fmt.Errorf("get ordering property max value: %w", err)
	}
}
```
</td>
</tr>
</table>

File: source/iterator/snapshot.go (hunk 7 above)
Comment (alyakimenko) [reply source/iterator/snapshot.go:91]: Yeah, for sure. Added.

File: source/iterator/snapshot.go (hunk 3 above)
Comment (maksenius) [reply source/iterator/snapshot.go:56]: Not needed. Thanks

File: source/iterator/snapshot.go (hunk 6 above)
Comment (alyakimenko) [reply source/iterator/snapshot.go:172]: Yeah, sounds great, I've added a new constant `metadataEntityLabelsField` for this purpose.

File: source/iterator/snapshot.go (hunk 5 above)
Comment (maksenius) [reply source/iterator/snapshot.go:96]: ok, I agree with you. Lets leave it

**Review by maksenius — APPROVED** [review]

**Pull request merged**